### ISSUE TITLE
More devs with merge rights?

### ISSUE THREAD

**Issue by Frost-wood** [CleverRaven/Cataclysm-DDA#12042]
Continued from: https://github.com/CleverRaven/Cataclysm-DDA/pull/12015 so it doesn't clutter the pr.

I'm curious as to what the process is to adding more devs with merge rights?

Is it me or does the pr's these days seem more advanced that the ones in the past and thus need more time to mergetest?  It also seems that more are coming in daily as well... might it be time to add more devs to help merge?

**Comment by DavidKeaton** [comment]
*raises hand* I'm interested in the talks about this,

**Comment by Coolthulhu** [comment]
Also interested. If mergetesting is such a time-consuming work, it could probably be split up.
I'm pretty sure finding competent volunteers within DDA community wouldn't be hard. I could volunteer myself.

And if people are worried about balance concerns due to differing opinions on how to develop a game, it could be solved by having some short merge guideline containing rules such as "let one of the [devs here] have look at anything balance-related before testing". Or even by having manual assignment on tasks, though this would slow things down.

**Comment by drbig** [comment]
More people on the mergefront would be better for all of us.

However there are two major concerns:

1. Making sure the changes fit in terms of balance/lore/vision of the game
2. Making sure the changes pass some bar in terms of code quality

With the above I'd prefer for people that have direct merge rights to be both competent in code and cdda history, and also somewhat dedicated (it's not a 'one off' kind of 'job').

Approaches we can mix and match:

1. Some people can focus on bug fixes
2. Some people can focus on new features
3. Some people can focus on merging related branches into one, making sure they work as advertised (e.g. hot fixing if necessary), and then pass such pre-tested branch to someone more experienced for final approval
4. Some people can focus on the 'works on platform X issues', e.g. 'platform maintainers'

My general point is that we already have insane pace of stuff coming in and inevitably breaking here and there, so with more people merging in we need to make it *saner*, not just *faster*, or we'll end up with even more mess than we started with.

**Comment by kevingranade** [comment]
I've invited both @Coolthulu and @BevapDin to the team, which means merge
rights.
BevapDin, if you don't want the responsibility for any reason, feel free to
refuse, you won't hurt my feelings, and I hope we can still be friends ;)
Coolthulu stuck his foot in it and volunteered already, so that seems set.
The basic tennets we follow, which you're probably aware of, but I'm
reiterating just in case, and I need to stick these somewhere in the repo
to make them more public...
1. Only do as much of a particular task as you want to, if you have some
feature or bugfix you really want to get in, work on that and don't worry
about the PR or issue backlogs.
2. Don't merge your own PRs unless they're an emergency patch to fix the
build or similar.

My apologies for putting this off for so long, I've been meaning to ask
BevapDin for months and just never got around to it, and Coolthulu has been
contributing at a level where I'm dumb to have not asked him for quite a
while.

Side note about bounties in case anyone is concerned, the funding party has
to confirm that the issue has been resolved in BountySource before any
funds are awarded, so Coolthulu doesn't have a 'give me money' button
sitting there tempting him to hit it.
To be doubly clear, the party in question is me in the case of the
kickstarter bounties.

**Comment by Coolthulhu** [comment]
Coolthulhu reporting in on the dev station!

I'm not sure how to handle balancing, so for now I'll stay with the technical stuff, simple cases like extra food and content additions that are mostly flavor. Unless someone PRs something objectively good, like buffing Deft or something like that.

As for the money (a valid concern) - me applying for a dev is partially motivated by it, but I want to claim the bounties the "legal" way. By helping with others' PRs so that other devs have more time to possibly review mine.
Plus, I kinda emotionally invested in the project. After working so much on it for free, there's no way I'd just run off with the money even if I could.

If someone has good PRs for a new dev to review, link/assign to me. For now all I see are debatable, WiP, balance-related, "claimed" or mine.

**Comment by mrskin** [comment]
Would it be at all helpful for there to be a workflow that someone like me can follow to get a PR closer to mergability? Some sort of peer review, to help the team members make their decision more quickly about merging something?

**Comment by drbig** [comment]
@mrskin Yeah I was thinking about it (and suggesting some approaches above).

First thing is testing your PR - you *should always first do it on your own* (does it build? works as intended? can I break/abuse it?). Fix all you see before you PR. But that should be obvious...

The less obvious factors are code quality, going in line with lore/vision, and making sure it works everywhere. And for that another pair of eyes would be helpful indeed. Such people could provide first feedback/patches before bothering anyone else.

I'm not arguing here for some hierarchy, but I think a general set of rules could be useful for everybody. I for one am ok with providing feedback on bug fixes and maintaining linux as a build platform.

**Comment by Frost-wood** [comment]
usually:
[WIP] prs that are still being worked on and are not ready for merging
[CR] code review
[Review] needs other devs to review the pr for viability to merge
[Ready] Ready for merging.
I found it best in my experience to make a pr when its done or almost done, the longer a pr sits the more likely it is to pick up merge conflicts.  

Json PR's are apparently easy to test and merge, so they get in faster.

Smaller PR's are eaiser to merge, and fix before you go all out on an idea that doesn't mesh with the vision of the game.  In a bigger PR you might have to scrap the entire pr and begin from scratch.

**Comment by Coolthulhu** [comment]
As for code quality, we have
https://github.com/CleverRaven/Cataclysm-DDA/blob/master/CONTRIBUTING.md
It does mention astyle, but it would be cool to have examples or a guideline. Not everyone knows astyle and not all of the arguments have unambiguous names.

Also, things like whether `const` should be before or after the type, should `const` be used in declarations where not necessary (`f( const int )`), whether `*` and `&` should have a space on the left or the right (or both).
This may be a trivial matter, but it's nice to have clear code with uniform style. Especially for people with obsession about pretty code.

**Comment by drbig** [comment]
@Coolthulhu Don't get me started on `astyle`... when was it last applied?

**Comment by Coolthulhu** [comment]
@drbig In some files probably more than 2 years. The problem with astyle is the number of conflicts it generates. If we had a good way or warning everyone that a file will be astyled soon, it could be done safely.
It would probably be a good idea to astyle the entire src (and maybe data/json, if jsons can be astyled) with stable releases.

**Comment by drbig** [comment]
@Coolthulhu Indeed in the freeze window before a release. Please... oh so much please... Then we can establish a simple rule: 'astyle before commit!' and be done with it forever.

**Comment by Coolthulhu** [comment]
Not sure about that astyle before commit thing. That would require all contributors to get astyle and learn how to use it. Git is already a large barrier that stops many people from contributing.
However 'astyle before merging' is a good suggestion.

**Comment by drbig** [comment]
@Coolthulhu Good points with the astyle requirement, and with the practical solution 'astyle before merging'. I could even see the latter as a 'duty' of merge-people (and it can be easily automated).

**Comment by drbig** [comment]
On the other hand 'get astyle and learn how to use it' may not be a *bad* thing. 'Git is already a large barrier that stops many people from contributing' - can't disagree. Still, with the amount of traffic here do we really need to put 'easy in' before 'good in'? (a general thought)

**Comment by Coolthulhu** [comment]
In this case astyle is more of a cosmetic thing. If it was something like requiring a valgrind run, then it would be a whole different matter. It's nice to have nice code, but it's not vital.
If we manage to get that full astyle at stable, it would probably be a good idea to ask contributors to use astyle (and do it for them if they forget), but I don't think it should be a requirement. Unless it gets automated, that is, but then the whole issue would be moot.

**Comment by i2amroy** [comment]
For topics of styling we've basically assumed a "style it when you change it" approach. It makes the files look pretty piecemeal at times, but with the amount of changes going in there isn't really a good time to just hit a file on the head with the style stick and change it all at once without breaking bunches of other things for other people.

>I'm not sure how to handle balancing, so for now I'll stay with the technical stuff, simple cases like extra food and content additions that are mostly flavor. Unless someone PRs something objectively good, like buffing Deft or something like that.

Feel free. Most of use start out in one area or another before spreading out a bit. Personally I was basically the exact opposite; *all* I merged for the first month or so was balance things and none of the technical ones. :P And when in doubt don't worry about asking the opinions of the other devs, there's a handful of us now to get decisions made if need be.

**Comment by drbig** [comment]
@i2amroy "style it when you change it" - been poked on that, tried my best, it's still a mess (and we do know it, and the reasons for it).

I've been suggesting it for some time, but I'd really like for us to make the decision to run astyle on the whole thing during next(...) release freeze. I know it will break PRs. It will probably generate grief. But it will be the good thing to do for the future.

And I shiver solely on the though of running cdda through valgrind... which may just as well make me try it.

**Comment by i2amroy** [comment]
Personally I'm fine with running a mass styling on everything (except the JSON's of course, we really need to write our own styler for those) come the next feature freeze, my post was mainly to reiterate the current view of the project. Assuming we publicize that we are going to be doing one ahead of time (and thus give fair warning to those who are planning PR's) I don't think there would be any huge problems with it.

**Comment by BevapDin** [comment]
BevapDin reporting in on the dev station! And I'm grateful for trust.

> BevapDin, if you don't want the responsibility for any reason, feel free to
refuse, you won't hurt my feelings, and I hope we can still be friends ;)

I'm not a fan of responsibility, but that won't stop me here. However, I probably not merging any content related stuff like recipes, item data and similar. That's just not my thing.

About the style, I usually apply style to the functions that I'm going to change heavily and commit that in a separate commit (usually the first in a PR). That works quite nice. Completely styling a file would be nice, but the many conflicts will be very annoying to contributors, it does not seem worth the trouble.

**Comment by Frost-wood** [comment]
congrats to the new devs with merge rights-tm

**Comment by vache** [comment]
Can I apply to be a dev with rejection rights?  I don't want to approve anything, just shoot things down.

**Comment by i2amroy** [comment]
>Can I apply to be a dev with rejection rights? I don't want to approve anything, just shoot things down.

No. I'm going to exercise my rejection rights to refuse your request. XD

**Comment by kevingranade** [comment]
Things you can do to help get PRs merged without merge rights:
Review the code, then point out potential issues with the code and/or
comment that you've reviewed it and it looks good to you (both happens when
your comments are nice-to-haves and you think it could be merged without
addressing them).
Merge the code locally, build it, and play around with it or formally test
it.  Then post the results, either good or bad.
If you do find a bug, feel free to supply pseudocode, a patch, or a pull
request that fixes it.
If extensive changes are required, you could even merge a PR locally, make
fixes, and then make that a separate PR.

**Comment by kevingranade** [comment]
Re: mass astyle.  I pushed a few astyle PRs a few months ago, targeting one
or two files at a time, but frankly I got sick of it in fairly short
order.  We can take another look at making a push in that direction if
there's interest.
I'm away from my computer for most of this weekend, can someone do a test
run and see just how bad it is?

**Comment by illi-kun** [comment]
This issue can be closed (at least for now), if (new) core devs don't feel they are overwhelmed by daily "merge-testing" activity.

**Comment by DavidKeaton** [comment]
I'd say this issue is currently resolved with Coolthulhu and BevapDin on the core dev team.

**Comment by kevingranade** [comment]
Indeed, it's resolved for the moment and I'll try and keep more on top of it in the future.

**Comment by DavidKeaton** [comment]
I mean I wouldn't mind being a dev, but you have added drbig too, so that help is nice!

**Comment by drbig** [comment]
@DavidKeaton I'm not merging stuff (= not a core dev). Unless there's an explicit permission. I try to keep tabs on issues (label check etc.).
